filebeat/input/filestream/copytruncate_prospector.go

```diff
@@ -0,0 +1,344 @@
+			r.table[original].rotated[idx].src = src
+			return idx
+		}
+	}
+
+	r.table[original].rotated = append(r.table[original].rotated, rotatedFileInfo{rotated, src, time.Time{}})
+	r.sorter.sort(r.table[original].rotated)
+
+	for idx, fi := range r.table[original].rotated {
+		if fi.path == rotated {
+			return idx
+		}
+	}
+
+	return -1
+}
+
+// addRotatedFile adds a new rotated file to the list and returns its index.
+// if a file is already added, the source is updated and the index is returned.
+func (r rotatedFilestreams) removeRotatedFile(original, rotated string) {
+	for idx, fi := range r.table[original].rotated {
+		if fi.path == rotated {
+			r.table[original].rotated = append(r.table[original].rotated[:idx], r.table[original].rotated[idx+1:]...)
+			return
+		}
+	}
+}
+
+type copyTruncateFileProspector struct {
+	fileProspector
+	rotatedSuffix *regexp.Regexp
+	rotatedFiles  *rotatedFilestreams
+}
+
+// Run starts the fileProspector which accepts FS events from a file watcher.
+func (p *copyTruncateFileProspector) Run(ctx input.Context, s loginp.StateMetadataUpdater, hg loginp.HarvesterGroup) {
+	log := ctx.Logger.With("prospector", copyTruncateProspectorDebugKey)
+	log.Debug("Starting prospector")
+	defer log.Debug("Prospector has stopped")
+
+	defer p.stopHarvesterGroup(log, hg)
+
+	var tg unison.MultiErrGroup
+
+	tg.Go(func() error {
+		p.filewatcher.Run(ctx.Cancelation)
+		return nil
+	})
+
+	tg.Go(func() error {
+		ignoreInactiveSince := getIgnoreSince(p.ignoreInactiveSince, ctx.Agent)
+
+		for ctx.Cancelation.Err() == nil {
+			fe := p.filewatcher.Event()
+
+			if fe.Op == loginp.OpDone {
+				return nil
+			}
+
+			src := p.identifier.GetSource(fe)
+			switch fe.Op {
+			case loginp.OpCreate, loginp.OpWrite:
+				if fe.Op == loginp.OpCreate {
+					log.Debugf("A new file %s has been found", fe.NewPath)
+
+					err := s.UpdateMetadata(src, fileMeta{Source: fe.NewPath, IdentifierName: p.identifier.Name()})
+					if err != nil {
+						log.Errorf("Failed to set cursor meta data of entry %s: %v", src.Name(), err)
+					}
+
+				} else if fe.Op == loginp.OpWrite {
+					log.Debugf("File %s has been updated", fe.NewPath)
+				}
+
+				if p.fileProspector.isFileIgnored(log, fe, ignoreInactiveSince) {
+					break
+				}
+
+				// check if the event belongs to a rotated file
+				if p.isRotated(fe) {
+					log.Debugf("File %s is rotated", fe.NewPath)
+
+					p.onRotatedFile(log, ctx, fe, src, hg)
+
+				} else {
+					log.Debugf("File %s is original", fe.NewPath)
+					// if file is original, add it to the bookeeper
+					p.rotatedFiles.addOriginalFile(fe.NewPath, src)
+
+					hg.Start(ctx, src)
+				}
+
+			case loginp.OpTruncate:
+				log.Debugf("File %s has been truncated", fe.NewPath)
+
+				s.ResetCursor(src, state{Offset: 0})
+				hg.Restart(ctx, src)
+
+			case loginp.OpDelete:
+				log.Debugf("File %s has been removed", fe.OldPath)
+
+				// if file is rotated, stop harvester and clean up state
+				if p.isRotated(fe) {
```

Comment:
Do we want to put this functionality behind a setting? Stopping a collected file early (or without timeout) means we will loose logs as we didn't reach EOF yet. This might be required if we don't have enough bandwidth, but we potentially might have a race condition here and the harvester is already close to EOF, we just need to give it an extra second (minute if output has an hiccup).